### PR TITLE
SR-1741: Add non-source files from source directories and root, ignore files from .gitignore

### PR DESCRIPTION
**Pull request by ViktorSimko**

### CODE REVIEW

File: Sources/Xcodeproj/generate().swift [hunk 1]
```diff
@@ -69,11 +71,27 @@ public func generate(
...
+    var extraFiles = try findNonSourceFiles(path: srcroot)
+    if let sourcesExtraFiles = try? findNonSourceFiles(path: srcroot.appending(component: "Sources"), recursively: true) {
+        extraFiles.append(contentsOf: sourcesExtraFiles)
+    }
+    if let testsExtraFiles = try? findNonSourceFiles(path: srcroot.appending(component: "Tests"), recursively: true) {
+        extraFiles.append(contentsOf: testsExtraFiles)
+    }
+
+    // Get the ignored files and exclude them
+    let ignoredFiles = try repositoryProvider.openCheckout(at: srcroot).getIgnoredFiles()
```
Comment (aciidb0mb3r) [inline Sources/Xcodeproj/generate().swift]: I think we need to handle the case when there is no repository at `srcroot`.

File: Sources/Xcodeproj/generate().swift [hunk 2]
```diff
@@ -181,3 +199,17 @@ func findDirectoryReferences(path: AbsolutePath) throws -> [AbsolutePath] {
...
+
+/// Finds the non-source files from `path` recursively
+func findNonSourceFiles(path: AbsolutePath, recursively: Bool = false) throws -> [AbsolutePath] {
+    let filesFromPath = try walk(path, recursively: recursively)
+
+    return filesFromPath.filter {
+        if !isFile($0) { return false }
+        if let `extension` = $0.extension {
+            if SupportedLanguageExtension.cFamilyExtensions.contains(`extension`) { return false }
+            if SupportedLanguageExtension.swiftExtensions.contains(`extension`) { return false }
```
Comment (aciidb0mb3r) [inline Sources/Xcodeproj/generate().swift]: We can use `validExtensions` here instead.

File: Sources/Xcodeproj/generate().swift [hunk 3]
```diff
@@ -181,3 +199,17 @@ func findDirectoryReferences(path: AbsolutePath) throws -> [AbsolutePath] {
         return isDirectory($0)
     })
 }
+
+/// Finds the non-source files from `path` recursively
```
Comment (aciidb0mb3r) [inline Sources/Xcodeproj/generate().swift]: Since recursively is a parameter, this should be reworded.

File: Sources/Xcodeproj/generate().swift [hunk 4]
```diff
@@ -69,11 +71,27 @@ public func generate(
     // references in the project.
     let extraDirs = try findDirectoryReferences(path: srcroot)
 
+    // Find non-source files in the source directories and root that should be added
+    // as a reference to the project.
+    var extraFiles = try findNonSourceFiles(path: srcroot)
+    if let sourcesExtraFiles = try? findNonSourceFiles(path: srcroot.appending(component: "Sources"), recursively: true) {
```
Comment (aciidb0mb3r) [inline Sources/Xcodeproj/generate().swift]: I think we don't need to use `try?` here.

File: Sources/Xcodeproj/generate().swift [hunk 5]
```diff
@@ -69,11 +71,27 @@ public func generate(
...
+        extraFiles.append(contentsOf: sourcesExtraFiles)
+    }
+    if let testsExtraFiles = try? findNonSourceFiles(path: srcroot.appending(component: "Tests"), recursively: true) {
+        extraFiles.append(contentsOf: testsExtraFiles)
+    }
+
+    // Get the ignored files and exclude them
+    let ignoredFiles = try repositoryProvider.openCheckout(at: srcroot).getIgnoredFiles()
+    extraFiles = extraFiles.filter {
```
Comment (aciidb0mb3r) [inline Sources/Xcodeproj/generate().swift]: This could just be `extraFiles = extraFiles.filter(ignoredFiles.contains)`.

File: Sources/SourceControl/GitRepository.swift [hunk 6]
```diff
@@ -391,6 +391,16 @@ public class GitRepository: Repository, WorkingCheckout {
         return localFileSystem.isDirectory(AbsolutePath(firstLine))
     }
 
+    public func getIgnoredFiles() -> [AbsolutePath] {
```
Comment (aciidb0mb3r) [inline Sources/SourceControl/GitRepository.swift]: This could have a big impact on performance because we will end up looking at each file in the .build directory, which is generally big. We should find a way to do this in a better way.

Comment (ViktorSimko) [reply Sources/SourceControl/GitRepository.swift]: Thanks for the review.
Yeah, this slipped my mind.
Maybe we could use `git check-ignore` on the files found by `findNonSourceFiles`.
Or should we also ignore the files if there isn't a repository for the package?

Comment (abertelrud) [reply Sources/SourceControl/GitRepository.swift]: Using `git check-ignore` sounds like a good idea.  And if there isn't a repository for the package, then by definition no files are ignored in a Git sense, so I don't think there's an issue there.

Comment (aciidb0mb3r) [reply Sources/SourceControl/GitRepository.swift]: ah I meant to say that we should make sure we don't throw when there is no repository.

File: Sources/Xcodeproj/pbxproj().swift [hunk 7]
```diff
@@ -294,7 +297,6 @@ func xcodeProject(
             let name = (sourcesGroup == nil ? groupName : target.name)
             let group = (sourcesGroup ?? parentGroup)
                 .addGroup(path: (path == "." ? "" : path), pathBase: .projectDir, name: name)
-
```
Comment (aciidb0mb3r) [inline Sources/Xcodeproj/pbxproj().swift]: Please avoid whitespace diffs.

File: Sources/Xcodeproj/generate().swift [hunk 8]
```diff
@@ -69,11 +71,25 @@ public func generate(
...
+    let sourcesExtraFiles = try findNonSourceFiles(path: srcroot.appending(component: "Sources"), recursively: true)
+    extraFiles.append(contentsOf: sourcesExtraFiles)
+    let testsExtraFiles = try findNonSourceFiles(path: srcroot.appending(component: "Tests"), recursively: true)
+    extraFiles.append(contentsOf: testsExtraFiles)
+
+    // Get the ignored files and exclude them
+    let ignoredFiles = try repositoryProvider.openCheckout(at: srcroot).getIgnoredFiles()
+    extraFiles = extraFiles.filter {
+        return !ignoredFiles.contains($0)
+    }
```
Comment (aciidb0mb3r) [inline Sources/Xcodeproj/generate().swift]: I think we can pass a closure to walk function that will avoid iterating the ignored directories, which is a little better for performance. We should also get some tests to prove this PR works.

File: Sources/SourceControl/GitRepository.swift [hunk 9]
```diff
@@ -391,6 +391,16 @@ public class GitRepository: Repository, WorkingCheckout {
...
+        return queue.sync {
+            let result = try? Process.checkNonZeroExit(
+                args: Git.tool, "-C", path.asString, "ls-files", "-o", "-i", "--exclude-standard").chomp()
+            return result?.split(separator: "\n").map(String.init).map {
+                return path.appending(RelativePath($0))
+            } ?? []
```
Comment (aciidb0mb3r) [inline Sources/SourceControl/GitRepository.swift]: We should not ignore the failures here and mark this method throwing.

**Comment by giginet** [comment]
How should we handle global `.gitignore`?

**Comment by ViktorSimko** [comment]
I've added tests for this.

**Comment by ViktorSimko** [comment]
@aciidb0mb3r Could you please review this?

File: Sources/Xcodeproj/generate().swift [hunk 10]
```diff
@@ -77,10 +79,32 @@ public func generate(
...
+    let testsDirectory = srcroot.appending(component: "Tests")
+    if localFileSystem.isDirectory(testsDirectory) {
+        let testsExtraFiles = try findNonSourceFiles(path: testsDirectory, recursively: true)
+        extraFiles.append(contentsOf: testsExtraFiles)
+    }
+
+    if try repositoryProvider.checkoutExists(at: srcroot) {
+        // Get the ignored files and exclude them
+        let repository = try repositoryProvider.openCheckout(at: srcroot)
+        extraFiles = try extraFiles.filter { try !repository.isIgnored($0) }
```
Comment (hartbit) [inline Sources/Xcodeproj/generate().swift]: Minor nitpick: Can you surround the closure with parentheses to match the informal project style? For reference, it follows the Rule of Kevin (“When a trailing closure argument is functional, use parentheses. When it is procedural, use braces.”)

**Comment by hartbit** [comment]
Minor stylistic comment, but this LGTM! @aciidb0mb3r what do you think?

File: Sources/SourceControl/GitRepository.swift [hunk 11]
```diff
@@ -92,6 +92,13 @@ public class GitRepositoryProvider: RepositoryProvider {
         }
     }
```
Comment (aciidb0mb3r) [inline Sources/SourceControl/GitRepository.swift]: We should land git related changes in a separate PR.

File: Sources/SourceControl/GitRepository.swift [hunk 12]
```diff
@@ -387,6 +397,22 @@ public class GitRepository: Repository, WorkingCheckout {
         return localFileSystem.isDirectory(AbsolutePath(firstLine))
     }
 
+    /// Returns true if the file at `path` is ignored by `git`
+    public func isIgnored(_ path: AbsolutePath) throws -> Bool {
```
Comment (aciidb0mb3r) [inline Sources/SourceControl/GitRepository.swift]: How about `isGitIgnored`?

File: Sources/Xcodeproj/generate().swift [hunk 13]
```diff
@@ -77,10 +79,32 @@ public func generate(
     // references in the project.
     let extraDirs = try findDirectoryReferences(path: srcroot)
 
+    // Find non-source files in the source directories and root that should be added
+    // as a reference to the project.
+    var extraFiles = try findNonSourceFiles(path: srcroot)
+
+    let sourcesDirectory = srcroot.appending(component: "Sources")
```
Comment (aciidb0mb3r) [inline Sources/Xcodeproj/generate().swift]: We should use the (root) package object to get this path as it maybe something else other than Sources.

File: Sources/Xcodeproj/generate().swift [hunk 14]
```diff
@@ -77,10 +79,32 @@ public func generate(
...
+    let testsDirectory = srcroot.appending(component: "Tests")
+    if localFileSystem.isDirectory(testsDirectory) {
+        let testsExtraFiles = try findNonSourceFiles(path: testsDirectory, recursively: true)
+        extraFiles.append(contentsOf: testsExtraFiles)
+    }
+
+    if try repositoryProvider.checkoutExists(at: srcroot) {
+        // Get the ignored files and exclude them
+        let repository = try repositoryProvider.openCheckout(at: srcroot)
+        extraFiles = try extraFiles.filter({ try !repository.isIgnored($0) })
```
Comment (aciidb0mb3r) [inline Sources/Xcodeproj/generate().swift]: This is going to shell out for every single extra file which will have a huge performance impact. Is there a way to pass multiple files in an invocation?

File: Sources/Xcodeproj/generate().swift [hunk 15]
```diff
@@ -77,10 +79,32 @@ public func generate(
     // references in the project.
     let extraDirs = try findDirectoryReferences(path: srcroot)
 
+    // Find non-source files in the source directories and root that should be added
```
Comment (aciidb0mb3r) [inline Sources/Xcodeproj/generate().swift]: We should factor out this entire diff into a new method.

**Comment by ViktorSimko** [comment]
@aciidb0mb3r  Thanks for the review, I've made the requested changes.

**Comment by hartbit** [comment]
@aciidb0mb3r I don’t agree with calling the function `isGitIgnored`: it’s a requirement on `Repository` and won’t make any sense if and when we provide an implementation for any other repository system (SVN, Mercurial). I think it should definitely be called `isIgnored`.

**Comment by aciidb0mb3r** [comment]
Ah, thats a good point. Maybe we should call it something else? isIgnored feels a bit vague but I can't think of something better either.

**Comment by ViktorSimko** [comment]
I can also agree with @hartbit .
For me, isIgnored would be expressing enough.

**Comment by hartbit** [comment]
`is` singular does feel wrong as it checks for multiple files thought: `areFilesIgnored(_:)` or `areFilesIgnored(at:)`?

**Comment by ViktorSimko** [comment]
Yeah, that's right.
`areFilesIgnored(at:)` seems to be better.

**Comment by ViktorSimko** [comment]
I've removed the Git related changes from this.

File: Sources/Xcodeproj/generate().swift [hunk 16]
```diff
@@ -223,3 +231,41 @@ func generateSchemes(
...
+
+    for target in package.targets {
+        let sourcesDirectory = target.sources.root
+        if localFileSystem.isDirectory(sourcesDirectory) {
+            let sourcesExtraFiles = try findNonSourceFiles(path: sourcesDirectory, recursively: true)
+            extraFiles.append(contentsOf: sourcesExtraFiles)
+        }
+    }
+
+    if let checkout = workingCheckout {
```
Comment (aciidb0mb3r) [inline Sources/Xcodeproj/generate().swift]: Hmm, maybe we shouldn't do this unless we have a working checkout because otherwise we will end up adding dot files like .DS_Store

Comment (ViktorSimko) [reply Sources/Xcodeproj/generate().swift]: Yeah, I can agree on this, also I think in most cases there will be a working checkout, so requiring it for this feature shouldn't cause many issues.

**Comment by aciidb0mb3r** [comment]
@swift-ci test with toolchain

File: Sources/Xcodeproj/generate().swift [hunk 17]
```diff
@@ -223,3 +231,39 @@ func generateSchemes(
...
+///   - path: The path of the directory to get the files from
+///   - recursively: Specifies if the directory at `path` should be searched recursively
+func findNonSourceFiles(path: AbsolutePath, recursively: Bool = false) throws -> [AbsolutePath] {
+    let filesFromPath = try walk(path, recursively: recursively)
+
+    return filesFromPath.filter({
+        if !isFile($0) { return false }
+        if let `extension` = $0.extension, SupportedLanguageExtension.validExtensions.contains(`extension`) {
+            return false
+        }
```
Comment (aciidb0mb3r) [inline Sources/Xcodeproj/generate().swift]: We should also exclude dot files here.

**Comment by aciidb0mb3r** [comment]
Ok, this looks good to me! Thanks a lot for all your work here!!

**Comment by aciidb0mb3r** [comment]
@swift-ci please smoke test

**Comment by ViktorSimko** [comment]
Thank you for reviewing my code and helping me!
